File: C# Fundamentals/C# OOP Advanced/07.Iterators_Comparators/Ex/P02_Collection/ListyIterator.cs
using System;
using System.Collections;
using System.Collections.Generic;

public class ListyIterator<T> : IEnumerable<T>
{
    private List<T> elements;

    private int index;

    public ListyIterator(params T[] elements)
    {
        this.elements = new List<T>(elements);
    }

    public void PrintAll()
    {
        if (this.elements.Count == 0)
        {
            throw new InvalidOperationException("Invalid Operation!");
        }
        Console.WriteLine(string.Join(" ", this.elements));
    }

    public void Print()
    {
        if (this.elements.Count == 0)
        {
            throw new InvalidOperationException("Invalid Operation!");
        }
        Console.WriteLine(this.elements[this.index]);
    }
    
    public bool Move()
    {
        if(this.index + 1 < this.elements.Count)
        {
            this.index++;
            return true;
        }
        return false;
    }
    public bool HasNext()
    {
        if(this.index + 1> this.elements.Count - 1)
        {
            return false;
        }
        return true;
    }

    public IEnumerator<T> GetEnumerator()
    {
        foreach (var element in this.elements)
        {
            yield return element;
        }
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return this.GetEnumerator();
    }
}

File: Java/01.Advanced/01.Advanced/17.IteratorsComparators/Ex/src/t_02_collection/ListyIterator.java
package t_02_collection;

import java.util.Iterator;
import java.util.List;

public class ListyIterator implements Iterable<String> {
    private int index;

    private List<String> collection;

    public ListyIterator(String... collection) {
        this.collection = List.of(collection);
    }

    public boolean move() {
        if (iterator().hasNext()) {
            index++;

            return true;
        }
        return false;
    }

    public void print() {
        if (collection == null || collection.size() < 1 || index == collection.size()) {
            System.out.println("Invalid Operation!");
        } else {
            System.out.println(collection.get(index));
        }
    }

    public void printAll() {
        if (collection == null) {
            System.out.println("Invalid Operation!");
        } else {
            System.out.println(String.join(" ", collection));
        }
    }

    @Override
    public Iterator<String> iterator() {

        return new Iterator<>() {
            @Override
            public boolean hasNext() {
                return index < collection.size() - 1;
            }

            @Override
            public String next() {
                return collection.get(index++);
            }
        };
    }
}
